### PR TITLE
 Audit fixes

### PR DESCRIPTION
**Pull request by 0xAshish**

### CODE REVIEW

File: contracts/staking/validatorShare/ValidatorShare.sol [hunk 1]
```diff
@@ -81,11 +82,12 @@ contract ValidatorShare is ValidatorShareStorage {
...
 
-    function sellVoucher() public {
+    function sellVoucher(uint256 expectedExchangeRate) public {
         uint256 share = balanceOf(msg.sender);
         require(share > 0, "Zero balance");
-        uint256 _amount = exchangeRate().mul(share).div(EXCHANGE_RATE_PRECISION);
-
+        uint256 rate = exchangeRate();
+        uint256 _amount = rate.mul(share).div(EXCHANGE_RATE_PRECISION);
+        require(rate >= expectedExchangeRate, "More slippage then expectedExchangeRate");
```
Comment (jdkanani) [inline contracts/staking/validatorShare/ValidatorShare.sol:90]: We should move to `89`. We should do the calculation after the require condition.

File: contracts/staking/validatorShare/ValidatorShare.sol [hunk 2]
```diff
@@ -217,12 +219,4 @@ contract ValidatorShare is ValidatorShareStorage {
         locked = true;
         return activeAmount;
     }
-
```
Comment (jdkanani) [inline contracts/staking/validatorShare/ValidatorShare.sol:220]: Please add test case for the same.

File: contracts/staking/validatorShare/ValidatorShare.sol (hunk 1 above)
Comment (0xAshish) [reply contracts/staking/validatorShare/ValidatorShare.sol:90]: ah yes

File: contracts/staking/validatorShare/ValidatorShare.sol (hunk 2 above)
Comment (0xAshish) [reply contracts/staking/validatorShare/ValidatorShare.sol:220]: yup adding

File: contracts/root/RootChain.sol [hunk 3]
```diff
@@ -16,38 +16,19 @@ contract RootChain is RootChainStorage, IRootChain {
...
-    {
-        (
-            address proposer,
-            uint256 start,
-            uint256 end,
-            bytes32 rootHash,
-            bytes32 accountHash,
-            uint256 _borChainID
-        ) = abi.decode(
-            data,
```
Comment (AyushKaul) [inline contracts/root/RootChain.sol:37]: this looks more readable

Comment (0xAshish) [reply contracts/root/RootChain.sol:37]: this is due to max length of the line. We can change it in the config.
:D bad idea for this PR.

File: contracts/root/depositManager/DepositManager.sol [hunk 4]
```diff
@@ -132,14 +105,13 @@ contract DepositManager is
...
     }
 
-    function depositERC20ForUser(address _token, address _user, uint256 _amount)
-        public
-    {
+    function depositERC20ForUser(
+        address _token,
+        address _user,
+        uint256 _amount
+    ) public {
```
Comment (jdkanani) [inline contracts/root/depositManager/DepositManager.sol:112]: This lint reversed!

File: contracts/staking/validatorShare/ValidatorShare.sol [hunk 5]
```diff
@@ -59,8 +59,9 @@ contract ValidatorShare is ValidatorShareStorage {
                 : withdrawPool.mul(EXCHANGE_RATE_PRECISION).div(_withdrawShares);
     }
 
-    function buyVoucher(uint256 _amount) public onlyWhenUnlocked {
+    function buyVoucher(uint256 _amount, uint256 expectedExchangeRate) public onlyWhenUnlocked {
```
Comment (atvanguard) [inline contracts/staking/validatorShare/ValidatorShare.sol:62]: it's better to take `minShareAmount` as an input parameter, then
```
require(share >= minShareAmount, "Insufficient amount to buy share");
```

File: contracts/staking/validatorShare/ValidatorShare.sol [hunk 6]
```diff
@@ -81,11 +82,12 @@ contract ValidatorShare is ValidatorShareStorage {
         logger.logStakeUpdate(validatorId);
     }
 
-    function sellVoucher() public {
+    function sellVoucher(uint256 expectedExchangeRate) public {
```
Comment (atvanguard) [inline contracts/staking/validatorShare/ValidatorShare.sol:85]: Same as above.

**Pull request merged**